server/pulp/server/webservices/views/consumer_groups.py

```diff
@@ -0,0 +1,373 @@
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :param consumer_group_id: A consumer group ID.
+        :type consumer_group_id: str
+        :raises: OperationPostponed
+        """
+        body = request.body_as_json
+        units = body.get('units')
+        options = body.get('options')
+        task = factory.consumer_group_manager().uninstall_content(consumer_group_id,
+                                                                  units, options)
+        raise pulp_exceptions.OperationPostponed(task)
+
+
+class ConsumerGroupBindingsView(View):
+    """
+    Views for repository binding to the group.
+    """
+    @auth_required(authorization.CREATE)
+    def post(self, request, consumer_group_id):
+        """
+        Create a bind association between the consumers belonging to the given
+        consumer group by id included in the URL path and a repo-distributor
+        specified in the POST body: {repo_id:<str>, distributor_id:<str>}.
+        Designed to be idempotent so only MissingResource is expected to
+        be raised by manager.
+
+        :param request: WSGI request object
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :param consumer_group_id: The consumer group ID to bind.
+        :type consumer_group_id: str
+        :raises: MissingResource if group id does not exist
+        :raises: InvalidValue some parameters are invalid
+        :raises: OperationPostponed
+        """
+        params = request.body_as_json
+        repo_id = params.get('repo_id')
+        distributor_id = params.get('distributor_id')
+        binding_config = params.get('binding_config', None)
+        options = params.get('options', {})
+        notify_agent = params.get('notify_agent', True)
+        missing_resources = verify_group_resources(consumer_group_id, repo_id, distributor_id)
+        if missing_resources:
+            if 'group_id' in missing_resources:
+                raise pulp_exceptions.MissingResource(**missing_resources)
+            else:
+                raise pulp_exceptions.InvalidValue(list(missing_resources))
+        bind_args_tuple = (consumer_group_id, repo_id, distributor_id, notify_agent,
+                           binding_config, options)
+        async_task = bind.apply_async(bind_args_tuple)
+        raise pulp_exceptions.OperationPostponed(async_task)
+
+
+class ConsumerGroupBindingView(View):
+    """
+    Represents a specific consumer group binding.
+    """
+    @auth_required(authorization.DELETE)
+    def delete(self, request, consumer_group_id, repo_id, distributor_id):
+        """
+        Delete a bind association between the consumers belonging to the specified
+        consumer group and repo-distributor. Designed to be idempotent.
+
+        :param request: WSGI request object
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :param consumer_group_id: A consumer group ID.
+        :type consumer_group_id: str
```

Comment:
On 01/27/2015 03:15 PM, bmbouter wrote:
> I don't think str is an actual python type.

str is a built-in Python type:

https://docs.python.org/2/library/functions.html#str

However, I usually list basestring as my accepted input types, if only
because most of the code I use would work the same on unicode and str types.